crates/bevy_ecs/src/entity/allocator.rs

```diff
@@ -0,0 +1,909 @@
+    unsafe fn get(&self, full_index: u32) -> Entity {
+        let (chunk, index, _) = self.index_in_chunk(full_index);
+        // SAFETY: Caller ensures this index was set
+        unsafe { chunk.get(index) }
+    }
+
+    /// Sets an entity at an index.
+    ///
+    /// # Safety
+    ///
+    /// This must not be called concurrently.
+    /// Access does not conflict with another [`Self::get`].
+    #[inline]
+    unsafe fn set(&self, full_index: u32, entity: Entity) {
+        let (chunk, index, chunk_capacity) = self.index_in_chunk(full_index);
+        // SAFETY: Ensured by caller and that the index is correct.
+        unsafe { chunk.set(index, entity, chunk_capacity) }
+    }
+
+    /// Iterates the entities in these indices.
+    ///
+    /// # Safety
+    ///
+    /// [`Self::set`] must have been called on these indices before to initialize memory.
+    #[inline]
+    unsafe fn iter(&self, indices: core::ops::RangeInclusive<u32>) -> FreeBufferIterator {
+        FreeBufferIterator {
+            buffer: self,
+            indices,
+            current: [].iter(),
+        }
+    }
+}
+
+impl Drop for FreeBuffer {
+    fn drop(&mut self) {
+        for index in 0..Self::NUM_CHUNKS {
+            let capacity = Self::capacity_of_chunk(index);
+            // SAFETY: we have `&mut` and the capacity is correct.
+            unsafe { self.0[index as usize].dealloc(capacity) };
+        }
+    }
+}
+
+/// An iterator over a [`FreeBuffer`].
+///
+/// # Safety
+///
+/// [`FreeBuffer::set`] must have been called on these indices before to initialize memory.
+struct FreeBufferIterator<'a> {
+    buffer: &'a FreeBuffer,
+    indices: core::ops::RangeInclusive<u32>,
+    current: core::slice::Iter<'a, Slot>,
+}
+
+impl<'a> Iterator for FreeBufferIterator<'a> {
+    type Item = Entity;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(found) = self.current.next() {
+            return Some(found.get_entity());
+        }
+
+        let next_index = self.indices.next()?;
+        let (chunk, index, chunk_capacity) = self.buffer.index_in_chunk(next_index);
+
+        // SAFETY: Assured by constructor
+        let slice = unsafe { chunk.get_slice(index, self.len() as u32 + 1, chunk_capacity) };
+        self.indices = (*self.indices.start() + slice.len() as u32 - 1)..=(*self.indices.end());
+
+        self.current = slice.iter();
+        Some(self.current.next()?.get_entity())
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.indices.end().saturating_sub(*self.indices.start()) as usize;
+        (len, Some(len))
+    }
+}
+
+impl<'a> ExactSizeIterator for FreeBufferIterator<'a> {}
+impl<'a> core::iter::FusedIterator for FreeBufferIterator<'a> {}
+
+/// This tracks the state of a [`FreeCount`], which has lots of information packed into it.
+///
+/// - The first 33 bits store a signed 33 bit integer. This behaves like a u33, but we define 1^33 as 0.
```

Comment:
1^33 is 1, did this mean 2^33? 2^33 does not fit in a u33. Did this mean 2^33-1 or 2^22 maybe?

Reply:
```suggestion
/// - The first 33 bits store a signed 33 bit integer. This behaves like a u33, but we define `1 << 33` as 0.
```
Oops. Another good catch. We could also do 2^32, but I think this is clearer that we keep 32 bits plus one to determine sine.